File: problems/LRU Cache/lruCache.cs
public class LRUCache {

    public LRUCache(int capacity) {
        _capacity = capacity;
        _store = new Dictionary<int, ListNode>();

        _head.Next = _tail;
        _tail.Prev = _head;
    }

    public int Get(int key) {
        if (!_store.ContainsKey(key)) {
            return -1;
        }

        var currNode = _store[key];

        removeFromList(currNode);
        insertOnTop(currNode);

        return currNode.Value;
    }

    public void Put(int key, int value) {
        if (_store.ContainsKey(key)) {
            var currNode = _store[key];

            removeFromList(currNode);
            insertOnTop(currNode);

            currNode.Value = value;
        } else {
            var newNode = new ListNode {Key = key, Value = value};

            insertOnTop(newNode);
            _store.Add(key, newNode);
        }

        if (_capacity < _store.Count) {
            _store.Remove(_tail.Prev.Key);

            _tail.Prev.Prev.Next = _tail;
            _tail.Prev = _tail.Prev.Prev;
        }
    }

    private void insertOnTop(ListNode node) {
        _head.Next.Prev = node;
        node.Next = _head.Next;
        node.Prev = _head;
        _head.Next = node;
    }

    private void removeFromList(ListNode node) {
        node.Prev.Next = node.Next;
        node.Next.Prev = node.Prev;
    }

    private int _capacity;
    private Dictionary<int, ListNode> _store;
    private ListNode _head = new ListNode();
    private ListNode _tail = new ListNode();

    private class ListNode {
        public int Key {get; set;}
        public int Value {get; set;}
        public ListNode Prev {get; set;}
        public ListNode Next {get; set;}
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.Get(key);
 * obj.Put(key,value);
 */

File: problems/LRU Cache/lruCache.java
class LRUCache {

    public LRUCache(int capacity) {
        size = capacity;
        keyToNodeMap = new HashMap<>();
        fakeHead = new Node();
        fakeTail = new Node();

        fakeHead.next = fakeTail;
        fakeTail.prev = fakeHead;
    }
    
    public int get(int key) {
        if (!keyToNodeMap.containsKey(key)) {
            return -1;
        }
        
        Node node = keyToNodeMap.get(key);

        remove(node);
        insertAfter(node, fakeHead);

        return node.value;
    }
    
    public void put(int key, int value) {
        if (keyToNodeMap.containsKey(key)) {
            Node node = keyToNodeMap.get(key);

            node.value = value;
            remove(node);
            insertAfter(node, fakeHead);
        } else {
            if (size == keyToNodeMap.size()) {
                keyToNodeMap.remove(fakeTail.prev.key);
                remove(fakeTail.prev);
            }

            Node node = new Node();

            node.key = key;
            node.value = value;
            insertAfter(node, fakeHead);
            keyToNodeMap.put(key, node);
        }
    }

    int size;
    HashMap<Integer, Node> keyToNodeMap;
    Node fakeHead;
    Node fakeTail;

    class Node {
        Node prev;
        Node next;
        int key;
        int value;
    }

    void insertAfter(Node node, Node head) {
        Node next = head.next;
        
        head.next = node;
        node.prev = head;
        node.next = next;
        next.prev = node;
    }

    void remove(Node node) {
        Node prev = node.prev;
        Node next = node.next;

        next.prev = prev;
        prev.next = next;
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */
